### PR TITLE
Add directive support to all diagrams by preprocessing

### LINKED ISSUE #4755: Remove directives from grammars.
### Description

- In render, `detectInit` is called to parse the directives using regex and we're sending that to `addDirective`
- Then, when parsing, the jison grammar parses the directives again, then `parseDirective` of the `diagramDB` is called, which calls `parseDirective` in `DirectiveUtils.ts`. The only thing extra being done there is `p.setWrap`. This also calls `addDirective`, possibly adding a duplicate directive to the array. 

### Steps to reproduce

Follow the flow of code. 

### Screenshots

_No response_

### Code Sample

_No response_

### Setup

_No response_

### Suggested Solutions

- Remove all the directive syntax from grammar
- Remove the parseDirective function
- Handle the `p.setWrap` somehow.
- When calling render, remove the frontmatter, directives and comments in the beginning itself.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by sidharthv96**
## :bookmark_tabs: Summary

Removes the directive grammar from the jison files.
Directives, frontmatter and comments will be preprocessed and removed, before parsing. 
Hence, all diagrams will now support directives.

Resolves #4755 

## :straight_ruler: Design Decisions

A preprocess step was added, which will extract metadata and remove comments, directives and frontmatter. 
wrap and title needed to be handled separately for legacy support.

### :clipboard: Tasks

Make sure you

- [x] :book: have read the [contribution guidelines](https://github.com/mermaid-js/mermaid/blob/develop/CONTRIBUTING.md)
- [x] :computer: have added necessary unit/e2e tests.
- [x] :notebook: have added documentation. Make sure [`MERMAID_RELEASE_VERSION`](https://github.com/mermaid-js/mermaid/blob/develop/packages/mermaid/src/docs/community/development.md#3-update-documentation) is used for all new features.
- [x] :bookmark: targeted `develop` branch

### CODE REVIEW

File: packages/mermaid/src/diagrams/git/gitGraphParser.spec
Comment (Yokozuna59) [inline packages/mermaid/src/diagrams/git/gitGraphParser.spec:1]: Why this file has been deleted? Shouldn't we just add `.js` or `.ts` extension? Or all test cases in this file are also included in `packages/mermaid/src/diagrams/git/gitGraphParserV2.spec.js`?

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts
Comment (Yokozuna59) [inline packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts:15]: I'm not familiar with this `wrap` option, what does it do? Is it necessary to have it in `init` function of the diagram definition? I also saw it in `c4`.

**Review by aloisklink — CHANGES_REQUESTED** [review]
Looks good! This makes everything a lot neater!

Some things to do before merging:
  - Does this branch need to be updated now that https://github.com/mermaid-js/mermaid/pull/4750 has been merged?
  - Please add `/** @deprecated */` TSDoc warnings to https://github.com/mermaid-js/mermaid/blob/066e0967de7e296febc93d17774c8c102d7ac84b/packages/mermaid/src/diagram-api/types.ts#L13 and https://github.com/mermaid-js/mermaid/blob/066e0967de7e296febc93d17774c8c102d7ac84b/packages/mermaid/src/diagram-api/types.ts#L49 explaining that the `_parseDirective()` function is now a no-op, and why it's a no-op (maybe with a link to this PR).
  - Is it worth holding back on merging this until v11? I don't think there are any **BREAKING CHANGES** (since we've just replaced `_parseDirective()` with a no-op), so it should be okay, but it's probably worth testing this PR with the `@mermaidjs/mermaid-zenuml` plugin, just to be on the safe side. Maybe it's also worth renaming this PR to **Preprocess directives (e.g. `%% {init: }}%%`) before parsing diagrams)**, just to make it a bit more clear what this PR does in the Mermaid v10.x.x release notes.
  - Does this mean that `%% {init: ... } %%` directives are now supported in sankey diagrams, which previously didn't support these? If yes, it's probably worth documenting this in the PR description.

**Review by Yokozuna59 — COMMENTED** [review]
This is a partial review, I'll look into remaining files (about 10) in few days since I don't have enough time right now, or you can carry on without my second review :)

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts
Comment (aloisklink) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts:15]: That's a good point. What does `wrap` even do?

For other diagrams, should we have something like:

```js
init: ({ wrap }) => {
  if (wrap) {
    // or maybe just print a warning?
    throw new Error("This diagram does not support %%{wrap}%%");
  }
},
```

File: packages/mermaid/src/diagrams/class/parser/classDiagram.jison
Comment (knsv) [inline packages/mermaid/src/diagrams/class/parser/classDiagram.jison:32]: ❤️

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js
Comment (knsv) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:2057]: Sometimes tests address multiple things, then it can be good to split it up. Wrap for instance is actually tested in another test (in this case) so it is safe to remove.

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts
Comment (Yokozuna59) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts:15]: I see, but does it has to be inside the DB?
Can't we just use `getConfig`?

File: packages/mermaid/src/Diagram.ts
Comment (sidharthv96) [reply packages/mermaid/src/Diagram.ts:81]: I've changed it to `Pick<DiagramMetadata, 'title'>`, we should not be passing config here as it'll be confusing right now on which config to use.

File: packages/mermaid/src/utils/sanitizeDirective.ts
Comment (sidharthv96) [reply packages/mermaid/src/utils/sanitizeDirective.ts:9]: Because `utils.ts` had this nifty line at the top, it didn't matter what the type was.
https://github.com/mermaid-js/mermaid/blob/6e51f8fd9840a788bcca1f2cd2aabd78bcf3b74a/packages/mermaid/src/utils.ts#L1

Changing it to unknown adds ~10 errors, which are not really relevant in our case as the args is basically an object | array, hence `any`.

image

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:2064]: Wrap has nothing to do with `should handle one actor, when theme is dark and logLevel is 1 DX1 (dfg1)`, and was most likely an artefact from a copied test. No need to complicate the unit test as we have other places to test wrap as Knut pointed out.

File: packages/mermaid/src/diagrams/git/gitGraphParser.spec
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/git/gitGraphParser.spec:1]: Added back, fixed some tests, skipped few.

File: packages/mermaid/src/utils/sanitizeDirective.ts
Comment (aloisklink) [reply packages/mermaid/src/utils/sanitizeDirective.ts:9]: > Because `utils.ts` had this nifty line at the top, it didn't matter what the type was
>
> https://github.com/mermaid-js/mermaid/blob/6e51f8fd9840a788bcca1f2cd2aabd78bcf3b74a/packages/mermaid/src/utils.ts#L1

:rofl: Oh well, `any` is fine then! `unknown` would be even better (especially for security-based code), but if we're not changing existing code, there's no need to change the typing.

File: packages/mermaid/src/Diagram.ts
Comment (aloisklink) [reply packages/mermaid/src/Diagram.ts]: > I've changed it to `Pick<DiagramMetadata, 'title'>`, we should not be passing config here as it'll be confusing right now on which config to use. 
>
> Maybe, `Omit<DiagramMetadata, 'config'>` can be used.

:rocket: Awesome! The `Pick<DiagramMetadata, 'title'>` seems perfect :) We can always change it to `Omit<DiagramMetadata, 'config'>` once/if we add more fields to the `DiagramMetadata` type in the future.

**Review by aloisklink — APPROVED** [review]
Looks good! Thanks for adding those `@deprecated` warnings in https://github.com/mermaid-js/mermaid/pull/4759/commits/18079a44bedb09eda7831aaacc5c3b263c97c227.

The only minor potential improvement is to rename and move around some of the tests in the `packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js` file.

---

One last thing, this PR adds support for `%% {init: {}}%%` directives in all diagrams, including diagrams like sankey diagrams that previously didn't support this.

E.g. try rendering the following diagram in the current `mermaid.live`, and the Mermaid Live preview for this PR:

````markdown
```mermaid
sankey-beta
  %% changing 'showValues' to true works to modify the diagram
  %%{init: {'sankey': {'showValues': false}}}%%
  %% source,target,value
  Electricity grid,Over generation / exports,104.453
  Electricity grid,Heating and cooling - homes,113.726
```
````

`%%{wrap}%%` also previously caused errors in diagrams, but now all diagrams support it (even if it does nothing in most diagrams).

To be honest, I like `%% {init: {} }%%` being accepted in all diagrams, since it keeps things consistent (even if it's deprecated behavior) :smile:. The `%%{wrap}%%` I'm okay with too, even if it's a no-op in most diagrams. It probably should just be documented in the PR description, so that we know that it's expected behavior.

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts
Comment (aloisklink) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.ts:15]: > Wrap makes the text wrap on messages for instance.

Ah, that makes sense!

FYI, there is a [`MermaidConfig.wrap`](https://mermaid.js.org/config/schema-docs/config.html#wrap) value, but the documentation for it is empty, which is why I got a bit confused!

However, [`MermaidConfig.c4.wrap`](https://mermaid.js.org/config/schema-docs/config.html#wrap-1), and [`MermaidConfig.sequence.wrap`](https://mermaid.js.org/config/schema-docs/config.html#wrap-2) are documented.

---

> I see, but does it has to be inside the DB?
> Can't we just use `getConfig`?

They're currently two separate values (but it's probably okay to refactor this in a future PR).

If somebody does:

```mermaid-example
%%{init: {'sequence': {'wrap': false}}}%%
%%{wrap}%%
sequenceDiagram
    Alice->>+John: Hello John, how are you? I am writing a verrrrrrrrrrrry long text.
```

```mermaid
%%{init: {'sequence': {'wrap': false}}}%%
%%{wrap}%%
sequenceDiagram
    Alice->>+John: Hello John, how are you? I am writing a verrrrrrrrrrrry long text.
```

You'd expect `getConfig().sequence.wrap === false`, because that's what the `%%{init: {'sequence': {'wrap': false}}}%%` bit says.

However, the `%%{wrap}%%` is part of the DB, so it would be `true` in the DB.

To be honest, the only main difference I see is that a website could set `siteConfig.secure = ["wrap"]` to block people from modifying the `wrap` config value in a diagram, then doing  `%%{init: {'sequence': {'wrap': true}}}%%` won't work, but `%%{wrap}%%` will work, but I don't really see why anybody would want to do that.

Anyway, if we do want to move `wrap` out of the DB and just keep it in the `config`, that should be separate PR to keep things simple :)

File: packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js
Comment (sidharthv96) [reply packages/mermaid/src/diagrams/sequence/sequenceDiagram.spec.js:2064]: I've removed that block of tests as they are redundant now.

**Pull request merged**
